### ISSUE TITLE
Repeated calls to uniform() produce non-uniform output

### ISSUE THREAD

**Issue by mdenil** [torch/cutorch#97]
Repeated calls to `:uniform(0, 1)` produce non-uniform output.  This only happens with CudaTensors and only across many calls to uniform, filling a large tensor with a single call seems okay.

```
-- this produces uniform output
-- local block = 10000

-- this does not
local block = 1

local r = torch.zeros(block):cuda()
for j = 1, 10000/block do
    r:uniform(0, 1)
    -- this is also broken
    -- r:rand(r:size())

    for i = 1, r:size(1) do
        print(r[i])
    end
end
```

**Comment by dominikgrewe** [comment]
Yes, you're right. The output seems to be "less uniform" when not generating a whole batch. It seems to get better the more numbers you generate but it's still not great.

I'm guessing that CUDA's random number generator isn't really designed for generating single numbers. To be efficient, you want to generate lots of them at a time. This is opposed to the Torch RNG where it doesn't make a difference whether you generate a single number N times, or N numbers in a batch (both in terms of performance and the result).

I don't think this is a problem in practice. If you only want to generate single numbers at a time, you can just use torch.uniform().

**Comment by dominikgrewe** [comment]
We did some more digging into this:

It turns out that the numbers generated by our RNG are only properly distributed when generating at least a few thousand numbers at a time (we used the dieharder tests to verify that). This is only since we changed from using cuRAND's host API to its device API. The host API always generates good numbers.

Some profiling suggests that the host API is always generating 16384 (64*256) random numbers at a time and then copies numbers from that pool into the output buffer. I can think of two ways we could do something similar:

1) Generate a pool of random bits and transform them into the right distributions on-demand (this is what the host API is doing afaict).
2) Have different pools for different distributions (eg. uniform and normal) and transform those numbers to the right distribution on-demand.

While the second option may obviously use more memory than the first, it would probably be easier to implement because we don't have to think about how to transform random bits into uniformly (or normally) distributed floats, for example.

**Comment by soumith** [comment]
this was documented in NVIDIA's CUDA docs, or I remember reading so.

Second option might end up using significantly more memory. If the overhead is less than say 10-25MB and it would save you tons of time, go for option 2. If the overhead of going for option (2) will hurt people with 2GB or 4GB cards (for example), then it is not great, and option (1) might be the way.

**Comment by dominikgrewe** [comment]
Storing 16384 floats takes up 64KB, so if we keep a set of uniform, normal and log-normal numbers around (the distributions that we currently rely on), we'll only use 192KB of extra memory. And we could allocate that lazily.
I think it makes sense to take advantage of cuRAND's capabilities to give us numbers from these distributions directly, rather than doing it ourselves.

**Comment by soumith** [comment]
okay, then (2) sounds good. I agree with you that we take advantage of cuRAND directly to give us these numbers.

**Comment by fidlej** [comment]
A log-normal Y can be obtained from normal X by: Y = exp(X).
So we can avoid storing log-normal numbers.

**Comment by jucor** [comment]
I'm worried of storing separate pools: it opens up the door to introducing correlations depending on the order in which you call the pools, in an application that uses different distributions. 

All the (little) theoretical guarantees that I'm aware of rely on having a good source of independent uniform numbers, then transform it them to get a stream of independent numbers regardless of the distribution of each number. 

I'm not saying the two pools is a bad idea, but I'm saying it is a very nonstandard one for which we have no guarantees whatsoever -- something which I'm always worried about when it comes to fundamental tools like this.

It is likely that behind-the-scene cuRAND's generation of normal numbers actually calls the uniform generator and transforms them on-demand. I do not know for sure, but that is the way I have seen it done in all RNG toolboxes I've encountered so far. We should be able to port the transformations from these classic generators (e.g. randomkit) without too much difficulty. 

I would suggest option (1). The not-too-big extra work it requires is worth the maintained guarantees. Also, we shouldn't forget to add enough information to the state variable that we can reproduce the pool exactly, for example:
* storing the pool in addition to the current uniform RNG state: that makes for a bit larger state, but is the simpler to store/restore.
* storing the state as it was before the last pool-filling,  the number of variates sampled from there, and the number of variates, left in the pool, so as to rebuild the pool at restore time. That makes the smallest information, but also the heavier state restoration.
